src/managers/upgrades.py

```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python3
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for computing upgrades related statuses."""
+
+import logging
+
+import charm_refresh
+import ops
+from data_platform_helpers.advanced_statuses.models import StatusObject
+from data_platform_helpers.advanced_statuses.protocol import ManagerStatusProtocol
+from data_platform_helpers.advanced_statuses.types import Scope
+
+from core.cluster import ClusterState
+from core.workload import WorkloadBase
+from statuses import CharmStatuses
+
+logger = logging.getLogger(__name__)
+
+
+class UpgradesManager(ManagerStatusProtocol):
+    """Manage upgrades statuses, but nothing else."""
+
+    name: str = "upgrades"
+
+    def __init__(
+        self, state: ClusterState, workload: WorkloadBase, refresh: charm_refresh.Machines
+    ):
+        self.state = state
+        self.workload = workload
+        self.refresh = refresh
+
+    def get_statuses(self, scope: Scope, recompute: bool = False) -> list[StatusObject]:
+        """Compute the upgrades-relevant statuses.
+
+        Advanced Statuses defines the status priority order per component. It is not possible to
+        have some statuses of a component more important and some other statuses of the same
+        component less important.
+
+        While `refresh.[app|unit]_status_higher_priority` must be of higher priority than any other
+        status, `refresh.unit_status_lower_priority()` should only be set it if there is no other
+        status at all. We achieve this by having `upgrades_manager` last in the component priority
+        order, but setting the field `approved_critical_component` to True if the refresh_status is
+        of higher priority than any other status.
+
+        This logic ignores all statuses set directly by the refresh lib, as of refresh v3.1.0.
+        For more information: see https://canonical-charm-refresh.readthedocs-hosted.com/latest/add-to-charm/status/
+        """
+        status_list: list[StatusObject] = []
+
+        if not self.refresh:
+            return [CharmStatuses.ACTIVE_IDLE.value]
+
+        if refresh_app_status := self.refresh.app_status_higher_priority:
+            try:
+                app_status = self._convert_ops_status_to_advanced_status(refresh_app_status)
+                status_list.append(app_status)
+            except ValueError as e:
+                logger.error(e)
```

Comment:
```suggestion
        if refresh_app_status := self.refresh.app_status_higher_priority:
            app_status = self._convert_ops_status_to_advanced_status(refresh_app_status)
            status_list.append(app_status)
```

In the current implementation, if the conversion fails, the higher priority refresh status will be overridden by another status

If the conversion fails, another status should not be set

IMO, it would be easiest to let the exception get raised without catching it. I believe it is safer (less likely to cause downtime and data loss) to let an uncaught exception get raised than it is to override a higher priority refresh status

Typically if there's an uncaught exception here I imagine it would get caught in testing before release, but even if this happens in production I believe it's safer to raise an uncaught exception than to override a higher priority refresh status

(same thing for `refresh.unit_status_higher_priority`)

Reply:
I've updated this method and removed the error handling. If the status returned from `refresh.app_status_higher_priority | refresh.unit_status_higher_priority | refresh.unit_status_lower_priority` is not a valid ops status, this will raise.

If - for whatever reason - it doesn't raise a `ValueError` in `charm-refresh`, it will raise an `AttributeError` in `_convert_ops_status_to_advanced_status` when we want to access the `ops_status.message`. There is also unit test coverage for this case, to ensure we do not override with other statuses in error cases.